File: Algorithm/python/BAEKJOON/CLASS 2/9012.py
N = int(input())
answer = ""
for _ in range(N):
    parenthesis_string = []
    for parenthesis in input():
        if parenthesis == "(":
            parenthesis_string.append(parenthesis)
        elif parenthesis == ")":
            if len(parenthesis_string) != 0 and parenthesis_string[-1] == "(":
                parenthesis_string.pop()
            else:
                parenthesis_string.append(parenthesis)
    if len(parenthesis_string) == 0:
        answer = "YES"
    else:
        answer = "NO"
    print(answer)

File: Algorithm/javascript/BAEKJOON/CLASS 2/9012.js
// TRY 1
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on("line", (line) => {
  if (!line) {
    rl.close();
  } else {
    input.push(line);
  }
}).on("close", () => {
  solution(input);
  process.exit();
});

function solution(input) {
  let N = Number(input[0]);
  for (let i = 1; i <= N; i++) {
    parenthesisString = [];
    for (parenthesis of input[i]) {
      if (parenthesis == "(") {
        parenthesisString.push(parenthesis);
      } else if (parenthesis == ")") {
        if (
          parenthesisString.length != 0 &&
          parenthesisString[parenthesisString.length - 1] == "("
        ) {
          parenthesisString.pop();
        } else {
          parenthesisString.push(parenthesis);
        }
      }
    }
    if (parenthesisString.length == 0) {
      console.log("YES");
    } else {
      console.log("NO");
    }
  }
}
